File: src/connectors/virustotal.js
import { BaseConnector } from './base.js';
import https from 'https';

export class VirusTotalConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'VirusTotal';
        this.apiKey = process.env.VIRUSTOTAL_API_KEY;
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        if (!this.apiKey) return { status: 'skipped', error: 'VIRUSTOTAL_API_KEY not set' };

        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];

            if (inputType === 'domain') {
                const data = await this._apiGet(`/api/v3/domains/${inputValue}`);
                if (data?.data?.attributes) {
                    const attrs = data.data.attributes;
                    entities.push({
                        type: 'domain',
                        data: {
                            name: inputValue,
                            registrar: attrs.registrar,
                            created_at: attrs.creation_date ? new Date(attrs.creation_date * 1000).toISOString() : null,
                            reputation: attrs.reputation,
                            malicious: attrs.last_analysis_stats?.malicious || 0,
                            suspicious: attrs.last_analysis_stats?.suspicious || 0,
                            harmless: attrs.last_analysis_stats?.harmless || 0,
                            categories: attrs.categories,
                            source: 'virustotal'
                        },
                        confidence: 0.95
                    });

                    if (attrs.last_dns_records) {
                        for (const rec of attrs.last_dns_records) {
                            if (rec.type === 'A' || rec.type === 'AAAA') {
                                entities.push({
                                    type: 'ip',
                                    data: { address: rec.value, source: 'virustotal_dns' },
                                    confidence: 0.9,
                                    asKnown: { type: 'ip', value: rec.value }
                                });
                            }
                            entities.push({
                                type: 'dns_record',
                                data: { type: rec.type, name: inputValue, value: rec.value, ttl: rec.ttl },
                                confidence: 0.9
                            });
                        }
                    }
                }

                const subs = await this._apiGet(`/api/v3/domains/${inputValue}/subdomains?limit=40`);
                if (subs?.data) {
                    for (const sub of subs.data) {
                        entities.push({
                            type: 'subdomain',
                            data: { name: sub.id, parent_domain: inputValue, source: 'virustotal' },
                            confidence: 0.9,
                            asKnown: { type: 'domain', value: sub.id }
                        });
                    }
                }
            }

            if (inputType === 'ip') {
                const data = await this._apiGet(`/api/v3/ip_addresses/${inputValue}`);
                if (data?.data?.attributes) {
                    const attrs = data.data.attributes;
                    entities.push({
                        type: 'ip',
                        data: {
                            address: inputValue,
                            asn: attrs.asn,
                            as_owner: attrs.as_owner,
                            country: attrs.country,
                            network: attrs.network,
                            reputation: attrs.reputation,
                            malicious: attrs.last_analysis_stats?.malicious || 0,
                            suspicious: attrs.last_analysis_stats?.suspicious || 0,
                            harmless: attrs.last_analysis_stats?.harmless || 0,
                            source: 'virustotal'
                        },
                        confidence: 0.95
                    });
                }

                const resolutions = await this._apiGet(`/api/v3/ip_addresses/${inputValue}/resolutions?limit=20`);
                if (resolutions?.data) {
                    for (const res of resolutions.data) {
                        const hostname = res.attributes?.host_name;
                        if (hostname) {
                            entities.push({
                                type: 'domain',
                                data: { name: hostname, source: 'virustotal_resolution', resolved_at: res.attributes?.date ? new Date(res.attributes.date * 1000).toISOString() : null },
                                confidence: 0.7,
                                asKnown: { type: 'domain', value: hostname }
                            });
                        }
                    }
                }
            }

            if (inputType === 'url') {
                const urlId = Buffer.from(inputValue).toString('base64url');
                const data = await this._apiGet(`/api/v3/urls/${urlId}`);
                if (data?.data?.attributes) {
                    const attrs = data.data.attributes;
                    entities.push({
                        type: 'url',
                        data: {
                            url: inputValue,
                            title: attrs.title,
                            final_url: attrs.last_final_url,
                            reputation: attrs.reputation,
                            malicious: attrs.last_analysis_stats?.malicious || 0,
                            suspicious: attrs.last_analysis_stats?.suspicious || 0,
                            source: 'virustotal'
                        },
                        confidence: 0.9
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence || 0.5);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _apiGet(path) {
        return new Promise((resolve, reject) => {
            const req = https.get({
                hostname: 'www.virustotal.com',
                path,
                headers: { 'x-apikey': this.apiKey, 'Accept': 'application/json' },
                timeout: 15000
            }, res => {
                if (res.statusCode === 404) { res.resume(); return resolve(null); }
                if (res.statusCode === 429) { res.resume(); return reject(new Error('VirusTotal rate limited')); }
                if (res.statusCode >= 400) { res.resume(); return reject(new Error(`VirusTotal HTTP ${res.statusCode}`)); }
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); } catch { resolve(null); }
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('VirusTotal timeout')); });
        });
    }

    async healthCheck() {
        return { ok: !!this.apiKey, version: 'v3', hasApiKey: !!this.apiKey };
    }
}

File: cyclops/connectors/virustotal.py
import os
import base64
import aiohttp
from cyclops.connectors.base import BaseConnector


class VirusTotalConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "VirusTotal"
        self.api_key = os.environ.get("VIRUSTOTAL_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        if not self.api_key:
            return {"status": "skipped", "error": "VIRUSTOTAL_API_KEY not set"}
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            entities = []
            headers = {"x-apikey": self.api_key, "Accept": "application/json"}
            async with aiohttp.ClientSession(headers=headers) as session:
                if input_type == "domain":
                    data = await self._get(session, f"/api/v3/domains/{input_value}")
                    if data and data.get("data", {}).get("attributes"):
                        a = data["data"]["attributes"]
                        entities.append({"type": "domain", "data": {"name": input_value, "registrar": a.get("registrar"),
                            "reputation": a.get("reputation"), "malicious": (a.get("last_analysis_stats") or {}).get("malicious", 0),
                            "categories": a.get("categories"), "source": "virustotal"}, "confidence": 0.95})
                        for rec in a.get("last_dns_records") or []:
                            entities.append({"type": "dns_record", "data": {"type": rec["type"], "name": input_value, "value": rec.get("value"), "source": "virustotal"}, "confidence": 0.9})
                            if rec["type"] in ("A", "AAAA"):
                                entities.append({"type": "ip", "data": {"address": rec["value"], "source": "virustotal_dns"}, "confidence": 0.9, "as_known": {"type": "ip", "value": rec["value"]}})
                    subs = await self._get(session, f"/api/v3/domains/{input_value}/subdomains?limit=40")
                    for sub in (subs or {}).get("data", []):
                        entities.append({"type": "subdomain", "data": {"name": sub["id"], "parent_domain": input_value, "source": "virustotal"}, "confidence": 0.9, "as_known": {"type": "domain", "value": sub["id"]}})
                elif input_type == "ip":
                    data = await self._get(session, f"/api/v3/ip_addresses/{input_value}")
                    if data and data.get("data", {}).get("attributes"):
                        a = data["data"]["attributes"]
                        entities.append({"type": "ip", "data": {"address": input_value, "asn": a.get("asn"), "as_owner": a.get("as_owner"),
                            "country": a.get("country"), "reputation": a.get("reputation"), "source": "virustotal"}, "confidence": 0.95})
                elif input_type == "url":
                    url_id = base64.urlsafe_b64encode(input_value.encode()).decode().rstrip("=")
                    data = await self._get(session, f"/api/v3/urls/{url_id}")
                    if data and data.get("data", {}).get("attributes"):
                        a = data["data"]["attributes"]
                        entities.append({"type": "url", "data": {"url": input_value, "title": a.get("title"), "reputation": a.get("reputation"),
                            "malicious": (a.get("last_analysis_stats") or {}).get("malicious", 0), "source": "virustotal"}, "confidence": 0.9})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def _get(self, session, path):
        async with session.get(f"https://www.virustotal.com{path}", timeout=aiohttp.ClientTimeout(total=15)) as resp:
            if resp.status in (404, 429) or resp.status >= 400:
                return None
            return await resp.json()

    async def health_check(self):
        return {"ok": bool(self.api_key), "hasApiKey": bool(self.api_key)}
